File: Assets/src/model/items/Item.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using UnityEngine;

[Serializable]
public class Item : IModifierProvider {
  public virtual string displayName =>
  /// get rid of the "Item" prefix
    Util.WithSpaces(GetType().Name.Substring(4));

  public virtual Inventory inventory { get; set; }

  [field:NonSerialized] /// controller uses this, but gameplay may as well
  public event Action OnDestroyed;
  public List<IItemMod> mods = new List<IItemMod>();
  public virtual IEnumerable<object> MyModifiers => mods;

  private int _stacks;
  public virtual int stacks {
    get => _stacks;
    set {
      if (value < 0) {
        throw new ArgumentException("Setting negative stack!" + this + " to " + value);
      }
      if (value > stacksMax) {
        value = stacksMax;
      }
      _stacks = value;
      if (_stacks == 0) {
        Destroy();
      }
    }
  }
  public virtual int stacksMax => 1;

  // An item is disjoint if it cannot be merged or accept new stacks from other instances of the same
  // item type.
  public virtual bool disjoint => false;

  public Item(int stacks) {
    this.stacks = stacks;
  }

  public Item() {
    this.stacks = disjoint ? stacksMax : 1;
  }

  // must take a parameter so ItemController can invoke this without erroring
  public virtual void Destroy(object unused = null) {
    OnDestroyed?.Invoke();
    if (inventory != null) {
      inventory.OnItemDestroyed?.Invoke(this);
      inventory.RemoveItem(this);
    }
  }

  public void Drop(Actor a) {
    if (inventory != null) {
      inventory.TryDropItem(a.floor, a.pos, this);
    }
  }

  internal virtual string GetStats() {
    return ObjectInfo.GetDescriptionFor(this);
  }

  public virtual List<MethodInfo> GetAvailableMethods(Player player) {
    var methods = new List<MethodInfo>() {
      // GetType().GetMethod("Destroy"),
      GetType().GetMethod("Drop")
    };
    if (this is IEdible edible) {
      methods.Add(GetType().GetMethod("Eat"));
    }
    if (this is IUsable usable) {
      if (this is EquippableItem i && !i.IsEquipped) {
        // EquippableItems can only be used while equipped 
      } else {
        methods.Add(GetType().GetMethod("Use"));
      }
    }
    return methods;
  }

  public bool CanStackWith(Item other) {
    if (disjoint || other.disjoint || GetType() != other.GetType()) {
      return false;
    }
    return StackingPredicate(other);
  }

  // only called when other.GetType() == this.GetType()
  protected virtual bool StackingPredicate(Item other) {
    return true;
  }
}

public interface IItemMod {
  string displayName { get; }
}

public static class ItemExtensions {
  public static string GetStatsFull(this Item item) {
    var text = item.GetStats() + "\n";
    if (item.mods.Any()) {
      text += $"<#FFFF00>{string.Join("\n", item.mods.Select(m => m.displayName))}</color>\n";
    }
    if (item is IWeapon w) {
      text += Util.DescribeDamageSpread(w.AttackSpread);
    }
    if (item.disjoint) {
      text += $"Uses: {item.stacks}/{item.stacksMax}.";
    }
    return text.Trim();
  }

  /// Merge as many stacks as possible from other into this one.
  /// Return true if the other stack is now empty.
  /// NOTE: it's the responsibility of the caller to then dispose of the IStackable!
  public static bool Merge<T>(this T s, T other) where T : Item {
    var spaceLeft = s.stacksMax - s.stacks;
    var stacksToAdd = UnityEngine.Mathf.Clamp(other.stacks, 0, spaceLeft);
    s.stacks += stacksToAdd;
    other.stacks -= stacksToAdd;
    return other.stacks == 0;
  }

  public static T Split<T>(this T item, int splitStack) where T : Item {
    if (item.disjoint) {
      return null;
    } else {
      splitStack = Mathf.Clamp(splitStack, 0, item.stacks);
      T split = item.GetType().GetConstructor(new Type[0]).Invoke(new object[0]) as T;
      split.stacks = splitStack;
      item.stacks = item.stacks - splitStack;
      return split;
    }
  }
}

File: web/src/model/Item.ts
import { EventEmitter } from '../core/EventEmitter';
import { GameModelRef } from './GameModelRef';
import type { EquipmentSlot } from './Equipment';
import type { Inventory } from './Inventory';
import type { Actor } from './Actor';
import type { Entity } from './Entity';
import type { Player } from './Player';
import type { Soil } from './Tile';

// ─── Symbol tags for item interfaces ───

export const WEAPON_TAG = Symbol.for('IWeapon');
export const DURABLE_TAG = Symbol.for('IDurable');
export const STACKABLE_TAG = Symbol.for('IStackable');
export const USABLE_TAG = Symbol.for('IUsable');
export const EDIBLE_TAG = Symbol.for('IEdible');
export const PLANTABLE_TAG = Symbol.for('IPlantable');
export const STICKY_TAG = Symbol.for('ISticky');
export const TARGETED_ACTION_TAG = Symbol.for('ITargetedAction');

// ─── Item interfaces ───

export interface IWeapon {
  readonly [WEAPON_TAG]: true;
  readonly attackSpread: [number, number];
}

export interface IDurable {
  readonly [DURABLE_TAG]: true;
  durability: number;
  readonly maxDurability: number;
}

export function reduceDurability(durable: IDurable): void {
  durable.durability--;
  if (durable.durability <= 0 && durable instanceof Item) {
    durable.Destroy();
  }
}

export function increaseDurability(durable: IDurable, amount = 1): void {
  durable.durability = Math.min(durable.durability + amount, durable.maxDurability);
}

export interface IStackable {
  readonly [STACKABLE_TAG]: true;
  stacks: number;
  readonly stacksMax: number;
}

export interface IConditionallyStackable extends IStackable {
  canStackWith(other: IConditionallyStackable): boolean;
}

/** Merge as many stacks as possible from other into this. Returns true if other is now empty. */
export function mergeStacks<T extends IStackable>(target: T, source: T): boolean {
  const spaceLeft = target.stacksMax - target.stacks;
  const stacksToAdd = Math.max(0, Math.min(source.stacks, spaceLeft));
  target.stacks += stacksToAdd;
  source.stacks -= stacksToAdd;
  return source.stacks === 0;
}

export interface IUsable {
  readonly [USABLE_TAG]: true;
  use(actor: Actor): void;
}

export interface IEdible {
  readonly [EDIBLE_TAG]: true;
  eat(actor: Actor): void;
}

export interface IPlantable {
  readonly [PLANTABLE_TAG]: true;
  plant(actor: Actor, soil: Soil): void;
}

/** Marker: cannot unequip or destroy once equipped. */
export interface ISticky {
  readonly [STICKY_TAG]: true;
}

/** Item that requires player to select a target before use (e.g. place on tile, charm enemy). */
export interface ITargetedAction {
  readonly [TARGETED_ACTION_TAG]: true;
  readonly targetedActionName: string;
  targets(player: Player): Entity[];
  performTargetedAction(player: Player, target: Entity): void;
}

// ─── Item base class ───

/**
 * Base class for all items.
 * NOT an Entity — items live inside Inventory/Equipment.
 * ItemOnGround wraps an Item as an Entity for floor placement.
 * Port of C# Item.cs.
 */
export class Item {
  private _inventory: Inventory | null = null;

  get inventory(): Inventory | null {
    return this._inventory;
  }

  set inventory(value: Inventory | null) {
    this._inventory = value;
  }

  readonly onDestroyed = new EventEmitter<[]>();

  get displayName(): string {
    // Strip "Item" prefix and add spaces before capitals
    const name = this.constructor.name;
    const stripped = name.startsWith('Item') ? name.substring(4) : name;
    return stripped.replace(/([A-Z])/g, ' $1').trim();
  }

  getStats(): string {
    return '';
  }

  getStatsFull(): string {
    let text = this.getStats() + '\n';
    if (WEAPON_TAG in this) {
      const w = this as unknown as IWeapon;
      const [min, max] = w.attackSpread;
      text += min === max ? `Damage: ${min}. ` : `Damage: ${min}-${max}. `;
    }
    if (DURABLE_TAG in this) {
      const d = this as unknown as IDurable;
      text += `Durability: ${d.durability}/${d.maxDurability}.`;
    }
    return text.trim();
  }

  Destroy(): void {
    if (this.inventory != null) {
      this.onDestroyed.emit();
      this.inventory.onItemDestroyed.emit(this);
      this.inventory.removeItem(this);
    }
  }

  Drop(actor: Actor): void {
    if (this.inventory != null && actor.floor) {
      this.inventory.tryDropItem(actor.floor, actor.pos, this);
    }
  }

  getAvailableMethods(): string[] {
    const methods: string[] = ['Drop'];
    if (EDIBLE_TAG in this) methods.push('Eat');
    if (USABLE_TAG in this) methods.push('Use');
    if (TARGETED_ACTION_TAG in this) {
      methods.push((this as unknown as ITargetedAction).targetedActionName);
    }
    return methods;
  }
}

// ─── EquippableItem ───

/**
 * Item that can be equipped in a specific slot.
 * Port of C# EquippableItem.cs.
 */
export abstract class EquippableItem extends Item {
  abstract get slot(): EquipmentSlot;

  protected get player() {
    return GameModelRef.main.player;
  }

  get isEquipped(): boolean {
    return this.inventory != null;
  }

  Equip(actor: Actor): void {
    (actor as any).equipment?.addItem(this);
  }

  Unequip(actor: Actor): void {
    (actor as any).inventory?.addItem(this);
  }

  /** Called when placed into equipment. Override for custom behavior. */
  OnEquipped(): void {}

  /** Called when removed from equipment. Override for custom behavior. */
  OnUnequipped(): void {}

  getAvailableMethods(): string[] {
    const methods = super.getAvailableMethods();
    // Check if in inventory (can equip) or in equipment (can unequip)
    const player = GameModelRef.mainOrNull?.player;
    if (player) {
      if (player.inventory.hasItem(this)) {
        methods.push('Equip');
      } else if (player.equipment.hasItem(this)) {
        methods.push('Unequip');
      }
      // Sticky items can't be unequipped/dropped/destroyed
      if (STICKY_TAG in this) {
        const idx = methods.indexOf('Unequip');
        if (idx >= 0) methods.splice(idx, 1);
        if (player.equipment.hasItem(this)) {
          const dropIdx = methods.indexOf('Drop');
          if (dropIdx >= 0) methods.splice(dropIdx, 1);
        }
      }
    }
    return methods;
  }
}
